[SwiftScan] Add SwiftScan APIs to replay the cache output

cachemeifyoucan opened the pull request:
Add new C APIs to libSwiftScan for cache querying and cache replays.

cachemeifyoucan commented:
I was trying to abstract out the compilation to avoid re-parsing command line, compute input/output and create instance in replay logics, but it turns out a lots of information needs to be passed, including:
* the diagnostics needed
* how to print diagnostics (color, format, etc.)
* inputs (to compute how to construct file specific diagnostic consumer)
* outputs (to replay)

So stick with replaying from command-line for now.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -497,6 +524,69 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t swiftscan_compute_cache_key(
+/// If the return value is SWIFTSCAN_CACHE_RESULT_ERROR, the error message is
+/// returned via \c error parameter and it caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_result_t swiftscan_cache_replay_compilation(
```
I was expecting this to have a "key" parameter and use that.  Instead it looks like it is triggering a full compilation if the result is not cached.

cachemeifyoucan replied:
swift compilation has more than one `key`. I have a patch for swift driver to compute all the keys already and driver/build system can query before calling replay. I actually would like this API to not compile at all, only replay. So preferably, only cache key goes into this API, but because all the limitation mentioned above, it is using command-line to reconstruct the keys and all the options for replay.

akyrtzi commented:
The libclang API has this kind of interaction with the build system:

1. key query API call: if the key exists it returns an abstract object that represents the outputs (the outputs are not downloaded from this call)
2. outputs materialize API call: downloads outputs locally
3. replay API call: receives arguments and the abstract object for outputs

Separating (1) from (2) provides flexibility on the client side for treating them as separate actions that can be recorded in the build log with their own timing info and executed potentially with different priority/scheduling. I would prefer to preserve that for swift as well instead of `swiftscan_cache_query` bundling both actions into the same API call.

And once an opaque object for outputs is introduced you can pass it to `swiftscan_cache_replay_compilation` to formalize that the outputs must already be found and potentially simplify the work that the replay function needs to do.

cachemeifyoucan commented:
> Separating (1) from (2) provides flexibility on the client side for treating them as separate actions that can be recorded in the build log with their own timing info and executed potentially with different priority/scheduling. I would prefer to preserve that for swift as well instead of swiftscan_cache_query bundling both actions into the same API call.

Maybe. But isn't that more round trip to remote that just adds latency? Maybe query can return faster without download, but what do you do if query succeed but materialize failed? Isn't that the same as a miss?

akyrtzi commented:
> But isn't that more round trip to remote that just adds latency?

Key/value queries are separated from CAS actions, both in the LLVMCAS APIs and the network protocols, there's no saving of latency in putting them behind one C API function instead of two.

> Maybe query can return faster without download, but what do you do if query succeed but materialize failed? Isn't that the same as a miss?

Yes, the client should treat it as a miss (or not, it is up to the client, maybe in a more strict mode it prefers to fail if the CAS download was not found, that's the benefit of the additional flexibility and control) and it will be more transparent to the client what exact action failed.

cachemeifyoucan commented:
Split `queryCacheKey` and `loadObject` into different API. Add a `reserved` field in replay function in case we want to add in the cache keys in the future.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please test

benlangmuir commented:
I'm a bit unclear on the intended design for the replay function. Currently it works by triggering a performFrontend that internally does the caching. IIUC that will replay all the cached compilations associated with that frontend job in the case of batch mode.  But if we have multiple keys for that and you have added one key parameter to the function for future use, won't you then need to make multiple calls to the same function?

I wonder if we should have a list of keys instead? I assume that even after you stop using performFrontend, you need to setup some state from the compiler arguments to do the replay.  That seems like something we want to avoid repeating multiple times for the same invocation if possible.

cachemeifyoucan commented:
> I'm a bit unclear on the intended design for the replay function. Currently it works by triggering a performFrontend that internally does the caching. IIUC that will replay all the cached compilations associated with that frontend job in the case of batch mode. But if we have multiple keys for that and you have added one key parameter to the function for future use, won't you then need to make multiple calls to the same function?
> 
> I wonder if we should have a list of keys instead? I assume that even after you stop using performFrontend, you need to setup some state from the compiler arguments to do the replay. That seems like something we want to avoid repeating multiple times for the same invocation if possible.

See my comment in the beginning. https://github.com/apple/swift/pull/68660#issuecomment-1728574118

The ideal situation is to replay from each key but the number of API required will grow significantly and might not be very stable. I was thinking if there are better way to communicate (like setup replay state from a JSON or a CAS object).

cachemeifyoucan commented:
After offline discussion, we are going to revise the cache key structure and this patch will be updated with the new design in mind later.

cachemeifyoucan commented:
Update the patch to using the new cache key schema and new C APIs.

Add a tool `swift-scan-test` that can be used to test libswiftscan C APIs.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan committed:
[Caching] Re-associate diagnostics cache key with InputFile

Change how cached diagnostics are stored inside the CAS. It used to be
stored as a standalone entry for a frontend invocation in the cache and
now it is switched to be associated with input files, stored together
with other outputs like object files, etc.

This enables cleaner Cache Replay APIs and future cached diagnostics
that can be splitted up by file contribution.

cachemeifyoucan committed:
[Caching] Add new CacheReplay APIs to libSwiftScan

Add new APIs libSwiftScan that can be used for cache query and cache
replay. This enables swift-driver or build system to query the cache and
replay the compilation results without invocation swift-frontend for
better scheduling.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
SwiftDriver side can be found at: https://github.com/apple/swift-driver/pull/1453

cachemeifyoucan commented:
ping

artemcm commented on tools/libSwiftScan/libSwiftScan.cpp:
```diff
@@ -57,11 +71,33 @@ class SwiftScanCAS {
   std::shared_ptr<llvm::cas::ActionCache> Cache;
 };
+
+struct SwiftCachedCompilation {
```
Would it be fair to call this a `SwiftCachedCompilationHandle`? Otherwise my first intuition is that this contains some kind of compilation result **state**.

artemcm commented on tools/libSwiftScan/libSwiftScan.cpp:
```diff
@@ -777,15 +882,618 @@ swiftscan_cache_compute_key(swiftscan_cas_t cas, int argc, const char **argv,
+  PrintingDiagnosticConsumer PDC(Err);
+  Inst.addDiagnosticConsumer(&PDC);
+
+  if (Invocation.getDiagnosticOptions().UseColor)
```
I guess in the long run it would be great to find a balance of how much of this we can factor out to something common as used by `performFrontend`.

cachemeifyoucan replied:
added FIXME

artemcm commented on tools/libSwiftScan/libSwiftScan.cpp:
```diff
@@ -777,15 +882,618 @@ swiftscan_cache_compute_key(swiftscan_cas_t cas, int argc, const char **argv,
+  auto *Instance = new SwiftScanReplayInstance();
+
+  // Drop "-frontend" if it is the first argument.
+  if (argc > 0 && llvm::StringRef(argv[0]) == "-fronkend") {
```
```suggestion
  if (argc > 0 && llvm::StringRef(argv[0]) == "-frontend") {
```

artemcm replied:
Does this actually break things, in practice?

cachemeifyoucan replied:
Mostly, not. Just for writing tests in `swift` will be really hard without hacks to drop `swift-frontend` and `-frontend` because you need to invoke swift with `%target-swift-frontend` and maybe I can do the hack in test tool instead of behind C APIs.

cachemeifyoucan replied:
Removed for now as I don't see this code being triggered.

cachemeifyoucan commented:
Also rewrite the patch to split Caching related code from DependencyScanning related code in libSwiftScan implementation so the file gets smaller and easier to read.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +539,130 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+    swiftscan_cached_compilation_t, void *ctx,
+    void (*callback)(void *ctx, swiftscan_cache_result_t,
+                     swiftscan_string_ref_t std_out,
+                     swiftscan_string_ref_t std_err,
```
The synchronous version (`swiftscan_cache_replay_compilation`) is not providing `std_out` and `std_err`.
I'd suggest to create an opaque object `swiftscan_cache_replay_result` that could encapsulate these kind of data.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,923 @@
+  delete unwrap(id);
+}
+
+bool swiftscan_cached_output_load(swiftscan_cas_t cas,
```
To be sure that all the remote network work will be done in these functions and the node will be fully materialized once the functions complete, both `swiftscan_cached_output_load` and `swiftscan_cached_output_load_async` should walk the object node and make sure any potential sub-nodes are also loaded.
Similar to what `clang_experimental_cas_loadObjectByString` and `clang_experimental_cas_loadObjectByString_async` are doing.

I'd suggest to move the code from  the `clang_experimental_cas_loadObject*` functions to an `ObjectStore` API, something like `loadNodeTree/loadNodeTreeAsync`, and use it here as well.

cachemeifyoucan replied:
hmm, I did check the clang function. It is only loading the direct children, which might not be enough. Should this be a full tree traversal?

akyrtzi replied:
The clang functions *are* doing a full tree traversal, did I do it wrong? Could you clarify why they are only doing direct children?

cachemeifyoucan replied:
My mistake. I probably read it wrong. But a full traversal is done in Swift API as well now.

akyrtzi commented on tools/swift-scan-test/swift-scan-test.cpp:
```diff
@@ -0,0 +1,256 @@
+
+} // namespace
+
+StringRef toString(swiftscan_string_ref_t str) {
```
You should mark this and the other functions in this file as `static`.

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +538,151 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_replay_compilation_async(
```
There's no compelling reason to have `swiftscan_cache_replay_compilation_async` variant, and the implementation of this function is not actually asynchronous, so I'd suggest to remove it.
If later on we have a compelling reason to add a real asynchronous version of `replay`, we can add it.

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +538,151 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// error, the error message is returned via \c error parameter and its caller
+/// needs to free the message using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_replay_result_t
+swiftscan_cache_replay_compilation(swiftscan_cas_t cas,
```
Does this need both `swiftscan_cas_t` and `swiftscan_cached_compilation_t`, could it be made that passing `swiftscan_cached_compilation_t` is sufficient?

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +538,151 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+    swiftscan_cache_cancellation_token_t *);
+
+/// Get return code from replay. Return code is 0 for success, and a non-zero
+/// value for failure.
```
To clarify, a "non-zero value for failure" means we cached a compilation that fails when replayed? Or does it mean that the replay failed for some reason? If it is the latter shouldn't it fail at the `swiftscan_cache_replay_compilation` call?

cachemeifyoucan replied:
My thinking is returnCode should be whatever the `swift-frontend` invocation should return. It is hard to match exactly but maybe I should document it to be like that?

akyrtzi replied:
So a "non-zero value" means we cached a compilation that fails when replayed? Which in practice will not happen because we won't cache a failed invocation?
I'd suggest to make it clear in the doc-comment this is the cached result and non-zero doesn't mean the replay failed.

cachemeifyoucan replied:
No, non-zero value absolutely means the replay failed. The returnCode matches the swift-frontend's return code if it is used for cache replay. Let me clarify that.

akyrtzi replied:
So if the replay failed where is the client getting the error description from, the `error` parameter of `swiftscan_cache_replay_compilation` or `swiftscan_cache_replay_result_get_stderr`?

cachemeifyoucan replied:
There is intentionally no error parameter. The replay is behaving like a process. It has a return code and it has stdout/error. It can be counted to have a nice diagnostics from stderr, but it is harder to implement a solution that the error message has to comeback from the parameter.

cachemeifyoucan replied:
Oh, I see, I actually have an error parameter here. Hmm... let me think about it.

cachemeifyoucan replied:
I see, with current implementation, the returnCode is not that useful. Let me get rid of it and using error parameter to indicate if there are any error.

akyrtzi reviewed (commented):
Could you add a `swiftscan_cached_output_get_casidstring` function, it will be useful for the client for logging purposes.

I also foresee a need for something like `swiftscan_cache_compilation_is_uncacheable`, for the case that the compiler decided that the compilation was uncacheable, so the client can distinguish this case. I'd suggest to add the function as a stub (always returning false) so clients can call it, and implement it properly in a future PR once we find a case for marking a key as 'uncacheable'.

cachemeifyoucan commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -441,6 +441,52 @@ typedef struct swiftscan_cas_options_s *swiftscan_cas_options_t;
+  SWIFTSCAN_OUTPUT_TYPE_BITSTREAM_OPT_RECORD = 19,
+  SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS = 20,
+  SWIFTSCAN_OUTPUT_TYPE_LAST = SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS
+} swiftscan_output_kind_t;
```
This allows swift-driver to perform the reverse mapping its `FileType`: https://github.com/apple/swift-driver/blob/eefc6581e50adf1ba6d492f4e023a824ab06dee8/Sources/SwiftDriver/SwiftScan/SwiftScanCAS.swift#L107

This allows swift-driver to associate the kind with the output `FileType` it knows about for the job it is trying to create. swift-driver also knows how to print `FileType` so I think this is better than just a name.

akyrtzi replied:
I'm still unclear why the kind enums are useful, doesn't the swift-driver already know what the outputs are from the invocation arguments? Why does it need to get the kinds from the cached compilation?

cachemeifyoucan replied:
For the enums, it allows it to tell which cached output is corresponding output from its computed job. It is probably not useful for current swift driver but if we need to convey the output kind, I prefer it to be a enum than a string. If we ever need the enum, it will render the string version useless. The downside is that we need a stable ordering for the enum value, which I don't see it being too big a problem.

akyrtzi replied:
You are creating an API that ties the client implementation with the compiler implementation, so it has a downside without providing any benefit currently.

If we have a function that just provides the name, the client can log it, and if later the compiler implementation adds a new output the client doesn't need to change anything in its code to be able to log the name of the new output.
But using an enum means we have to coordinate and sync-up both the client and the compiler in order to be able to log the name of a new output.

We can always add the kind back in later, when there is a benefit so the tradeoff of tying the implementations is worth it.

cachemeifyoucan replied:
The outputs needed from the compilation is first computed by driver, which send to frontend as command-line argument. It is currently tied together already and frontend should not try to produce output that are not expected by driver.

cachemeifyoucan commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +538,151 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// error, the error message is returned via \c error parameter and its caller
+/// needs to free the message using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC bool
+swiftscan_cached_output_load(swiftscan_cas_t cas, swiftscan_cached_output_t,
```
Sure. I don't mind the separation between a simple output object and a CAS but I do get the benefit of simpler APIs.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +539,130 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// returned via \c error parameter and its caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_replay_instance_t
+swiftscan_cache_replay_instance_create(int argc, const char **argv,
```
I don't see that being necessary for current version. `-working-directory` is an option that doesn't do anything for `swift-frontend`. Driver is supposed to expand all paths so frontend doesn't need to deal with it. However that is not a documented behavior but I like it that way.

benlangmuir replied:
> Driver is supposed to expand all paths so frontend doesn't need to deal with it.

SGTM.

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,838 @@
+        swift::c_string_utils::create_clone(toString(std::move(E)).c_str());
+    return false;
+  };
+  if (auto Err = Out->CAS.getProxy(Out->Ref).moveInto(Out->LoadedProxy))
```
This and `traverseAllCASNodes` should use `getProxyIfExists`, so the caller can distinguish between an error and "not found".
(I realized the clang side is doing the same mistake and I need to update it).

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,838 @@
+  llvm::raw_pwrite_stream &StdOut;
+};
+
+llvm::Error replayCompilation(SwiftScanReplayInstance &Instance,
```
Even though this is in an anonymous namespace, llvm style recommends to still use `static` for top-level functions for clarity.

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,838 @@
+  }
+
+  return llvm::Error::success();
+  ;
```
Leftover semicolon.

cachemeifyoucan commented on lib/DriverTool/swift_cache_tool_main.cpp:
```diff
@@ -367,10 +369,21 @@ int SwiftCacheToolInvocation::validateOutputs() {
   Instance.getDiags().addConsumer(PDC);
 
+  auto lookupFailed = [&](StringRef Key) {
+    llvm::errs() << "failed to find output for cache key " << Key << "\n";
```
This is a top level tool. It eventually needs to provide a user readable message and each action just prints its own error message as it has more context than the top level main function.

cachemeifyoucan commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,838 @@
+    return failure(createObjectNotFoundForUploadError());
+
+  Cache.putAsync(ID, **Result, /*globally=*/true,
+                 [&](llvm::Error E) { return failure(std::move(E)); });
```
error is always set. Note it is not a pointer but a struct. The state that is not set is the default state (a ref with length 0 and data to be nullptr).

cachemeifyoucan replied:
I forgot the call back here :( Good catch.

cachemeifyoucan commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +509,145 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// using `swiftscan_cache_cancellation_token_dispose`. If no token is needed,
+/// nullptr can be passed and no token will be returned.
+SWIFTSCAN_PUBLIC void swiftscan_cached_compilation_make_global_async(
+    swiftscan_cas_t cas, swiftscan_cached_compilation_t, void *ctx,
```
Actually, cached compilation doesn't know anything about Cache because it is not needed to read cached compilation and outputs. I can add that to make API cleaner.

akyrtzi replied:
Ok, keeping as is makes sense to me 👍

cachemeifyoucan commented:
@swift-ci please smoke test

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -492,6 +509,145 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+/// Get the output kind name for the given index in the cached compilation. The
+/// returned name needs to be freed by `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+swiftscan_cached_compilation_get_output_kind_name(
```
Sorry I wasn't clear, I meant instead of passing `swiftscan_cached_compilation_t` and and index:
```
swiftscan_string_ref_t
swiftscan_cached_compilation_get_output_name(swiftscan_cached_compilation_t, unsigned idx);
```

change to just passing a `swiftscan_cached_output_t`:
```
swiftscan_string_ref_t
swiftscan_cached_output_get_name(swiftscan_cached_output_t);
```

cachemeifyoucan replied:
I see. But the kind is stored in the compilation node, even it is implementation detail. Switching to the new API requires me to re-arrange the underlying container again ....

akyrtzi replied:
Ok, it's not a big deal, keep as is 👍

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,902 @@
+          ++NumPending;
+      }
+      if (!Inserted) {
+        finishedNode();
```
Oops, that is my mistake 😅but this line is wrong, it should be
```
      if (!Inserted) {
        return;
      }
```
without calling `finishedNode();`.

(I need to update the clang side..)

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,902 @@
+        swift::c_string_utils::create_clone(toString(std::move(E)).c_str());
+    return false;
+  };
+  if (auto Err = Out->CAS.getProxy(Out->Ref).moveInto(Out->LoadedProxy))
```
Should also use `getProxyIfExists` here.

akyrtzi commented on tools/libSwiftScan/SwiftCaching.cpp:
```diff
@@ -0,0 +1,902 @@
+      return failure(Proxy.takeError());
+
+    if (!*Proxy)
+      return false;
```
Should you do
```
*error = swift::c_string_utils::create_null();
```
here ?

I've seen this in other places, should the `error` parameter always be set by the functions?

cachemeifyoucan commented:
@swift-ci please smoke test

akyrtzi reviewed (approved):
Thank you for addressing all the feedback!

cachemeifyoucan commented:
@swift-ci please smoke test windows platform
